service/history/chasm/components/activity/handler.go

```diff
@@ -0,0 +1,160 @@
+}
+
+type ScheduleRequest struct {
+	Input []byte
+}
+type ScheduleResponse struct{}
+
+type RecordStartedRequest struct {
+	RefToken []byte
+}
+
+type RecordStartedResponse struct {
+	RefToken []byte
+	Input    []byte
+}
+
+type RecordCompletedRequest struct {
+	RefToken []byte
+	Output   []byte
+}
+
+type RecordCompletedResponse struct{}
+
+type DescribeActivityRequest struct{}
+
+type DescribeActivityResponse struct {
+	IsAbandonded  bool
+	StartedTime   time.Time
+	CompletedTime time.Time
+}
+
+func (h *ActivityHandler) NewActivity(
+	ctx context.Context,
+	request *NewActivityRequest,
+) (*NewActivityResponse, error) {
+	resp, activityRef, err := chasm.NewEntity(
+		ctx,
+		chasm.EntityKey{
+			NamespaceID: "default",
+			BusinessID:  "memo",
+			// in V1 we probably don't support specifying instanceID,
+			// need to change persistence implementation for supporting that.
+			// InstanceID:  uuid.New().String(),
+		},
+		NewScheduledActivity,
+		request,
+		chasm.WithBusinessIDPolicy(
+			chasm.BusinessIDReusePolicyAllowDuplicate,
+			chasm.BusinessIDConflictPolicyFail,
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	resp.RefToken = activityRef.Serialize()
+	return resp, err
+}
+
+func (h *ActivityHandler) RecordStarted(
+	ctx context.Context,
+	request *RecordStartedRequest,
+) (*RecordStartedResponse, error) {
+	// resp := &RecordStartedResponse{}
+
+	ref, err := chasm.DeserializeComponentRef(request.RefToken)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, startedActivityRef, err := chasm.UpdateComponent(
+		ctx,
+		ref,
+		(*Activity).RecordStarted,
+		request,
+		// chasm.EngineEagerLoadOption([]chasm.ComponentPath{
+		// 	{"Input"},
+		// }),
+	)
+
+	resp.RefToken = startedActivityRef.Serialize()
+	return resp, err
+}
+
+func (h *ActivityHandler) RecordCompleted(
+	ctx context.Context,
+	request *RecordCompletedRequest,
+) (*RecordCompletedResponse, error) {
+	ref, err := chasm.DeserializeComponentRef(request.RefToken)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, _, err := chasm.UpdateComponent(
+		ctx,
+		ref,
+		(*Activity).RecordCompleted,
+		request,
+	)
+
+	return resp, err
+}
+
+type GetActivityResultRequest struct {
+	RefToken []byte
+}
+
+type GetActivityResultResponse struct {
+	Output []byte
+}
+
+func (h *ActivityHandler) GetActivityResult(
+	ctx context.Context,
+	request *GetActivityResultRequest,
+) (*GetActivityResultResponse, error) {
+	ref, err := chasm.DeserializeComponentRef(request.RefToken)
+	if err != nil {
+		return nil, err
+	}
+
+	var resp *GetActivityResultResponse
+	if resp, _, err = chasm.PollComponent(
+		ctx,
+		ref,
+		func(a *Activity, ctx chasm.Context, _ *GetActivityResultRequest) bool {
+			return a.LifecycleState() == chasm.LifecycleStateCompleted
+		},
+		func(a *Activity, ctx chasm.MutableContext, _ *GetActivityResultRequest) (*GetActivityResultResponse, error) {
+			outputPayload, err := a.Output.Get(ctx)
+			resp.Output = outputPayload.Data
+			return resp, err
+		},
```

Comment:
I wonder if we'd be better off merging these two like so:

```suggestion
		func(a *Activity, ctx chasm.MutableContext, _ *GetActivityResultRequest) (*GetActivityResultResponse, error) {
			if a.LifecycleState() != chasm.LifecycleStateCompleted {
				return nil, chasm.ErrNoReady / *name TBD */
			}
			outputPayload, err := a.Output.Get(ctx)
			resp.Output = outputPayload.Data
			return resp, err
		},
```

Reply:
I think I prefer them separate, because what happens if you mutate something and then say "not ready"? That would be some weird violation that shouldn't be possible, and separate contexts enforces that at the type level.

If you need context from one to the other, make the ready function return `(bool, T)` and `T` is passed into the mutator

Reply:
I changed the signature of the validation to return (T, bool, error), though that means the mutator is now taking 4 params...

Reply:
Not sure why there needs to be a mutator in poll or why you need both an error and a bool.